### PR TITLE
Add support for attributes to the custom Thrive bbcode

### LINKED ISSUE #2325: Allow overriding compound name in compound bbcode tag for translations
In Turkish grammar; if last letter of a word is a strong consonant, it usually needs to be changed into a soft consonant depending on the suffix added. This is called "Final-obstruent devoicing". I am not quite familiar with bbcode but I assume it uses fixed words which may be a problem with the kind of words mentioned above. If otherwise, please ignore this.

To give an example of the situation, take the ammonia bbcode:

[thrive:compound]ammonia[/thrive:compound]

I assume this always shows up "Amonyak" in bbcode. There are situations I need to add a suffix to write it like this "Amonyağa". As "k" is a strong consonant, it should turn into "ğ" instead. However bbcode has the "k" hardcoded (?) in the word so I can only make it look like this: "Amonyaka". This makes the grammar incorrect.

### PR DESCRIPTION
**Pull request by Kasterisk**
**Brief Description of What This PR Does**

This adds support for specifying attributes in a custom Thrive bbcode tag. Currently only the `compound` tag having one for overriding the compound name, which could be added with the following syntax: `[thrive:compound text="Other name"]` (value must be enclosed within quotation marks for it to be parsed correctly). Useful for special cases in translations.

**Related Issues**

Closes #2325 

**Progress Checklist**

Note: before starting this checklist the issue should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author.

### CODE REVIEW

**Comment by hhyyrylainen** [comment]
I don't think this can make it into 0.5.4 because all translators would need to work super fast to finish updated translations today...

**Comment by Kasterisk** [comment]
Internally I think this is good to have in 0.5.4 (because this only adds extra functionality), the visible translations change could be reserved for the next release if no translators are doing it before 0.5.4.

**Comment by hhyyrylainen** [comment]
But this would not add any benefit for 0.5.4, right? So to be extra safe I'd like to postpone merging this by 1 day (If this is good to merge tomorrow, I can merge this just after making the release).

**Comment by Kasterisk** [comment]
That's fine by me.

File: src/general/StringUtils.cs
Comment (hhyyrylainen) [inline src/general/StringUtils.cs:51]: I *think* we may need a way to escape `"`'s within a string so you could have text like `[thrive:compound text="This is some \"stuff\" in the game"]` And of course double `\` would mean then a single `\`.

File: src/gui_common/CustomRichTextLabel.cs
Comment (hhyyrylainen) [inline src/gui_common/CustomRichTextLabel.cs:186]: I think naming wise this could be attributes to follow the convention set in html (If I understood this part correctly)

Comment (Kasterisk) [reply src/gui_common/CustomRichTextLabel.cs:186]: It's indeed called attributes in some [sources](https://s9etextformatter.readthedocs.io/Plugins/BBCodes/Custom_BBCode_syntax/) I found.

File: src/general/StringUtils.cs
Comment (Kasterisk) [reply src/general/StringUtils.cs:52]: Added.

File: src/general/StringUtils.cs
Comment (hhyyrylainen) [inline src/general/StringUtils.cs:91]: Why not return the list here instead of making an extra array allocation here?

File: src/general/StringUtils.cs
Comment (hhyyrylainen) [inline src/general/StringUtils.cs:51]:
```suggestion
    /// <param name="ignoreWithinQuotes">Ignore whitespace within quotation marks.</param>
```

File: src/general/StringUtils.cs
Comment (hhyyrylainen) [inline src/general/StringUtils.cs:82]: To me it looks like this is unusable method here https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regex.unescape?view=net-5.0 it seems to do much else than is required.
Basically a custom helper method that *only* does basic index for the substring changing if `\` is found here.
Also the substr call should have the adjusted cut location for escaping to save on one string instance creation.

File: src/gui_common/CustomRichTextLabel.cs
Comment (hhyyrylainen) [inline src/gui_common/CustomRichTextLabel.cs:147]: I think this split should only be done after the if on line 150

File: src/gui_common/CustomRichTextLabel.cs
Comment (hhyyrylainen) [inline src/gui_common/CustomRichTextLabel.cs:244]: Why not merge the sequential checks here?

File: src/gui_common/CustomRichTextLabel.cs
Comment (hhyyrylainen) [inline src/gui_common/CustomRichTextLabel.cs:249]: I think this would be nice to split to a general parsing method to convert list of "key=value" pairs into a Dictionary<string,string>.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: src/general/StringUtils.cs
Comment (Kasterisk) [reply src/general/StringUtils.cs:82]: In my opinion the simplest alternative could be calling string `Replace` after `input.Substr` for replacing occurrences of `\"` with `"`, this makes sense because `SplitByWhiteSpace` method only handle the custom escaped double quotes and nothing else.

Comment (hhyyrylainen) [reply src/general/StringUtils.cs:82]: That wouldn't handle `"text here \\"` that would be parsed incorrectly. This needs a proper parser with state tracking if next character should be ignored.

Comment (Kasterisk) [reply src/general/StringUtils.cs:82]: That would require multiple backslashes (more than two) in the source string to have the double backslashes on there (which is not usual in normal use-case). `StringUtils.SplitByWhiteSpace("some text \\", true).ForEach(e => GD.Print(e));` would output `some`, `text` and `\`. The escape sequence would have been handled by the compiler/gettext first before it is given to the method, the use of double quotes escape sequence in the method is the only special case where it requires a multi-backslash to properly have `\"` string constructed.

Comment (hhyyrylainen) [reply src/general/StringUtils.cs:82]: If someone literally writes `\\` on the translations website. That will get escaped so that when the game loads the text it will contain `\\` meaning that our parsing needs to handle that correctly.
In the previous engine I had a very complicated custom written string parsing library with a bunch of tests for edge cases like this. It had things like get the next string in quotes and it would handle nesting of ' and " quotes with escape character handling.

Comment (Kasterisk) [reply src/general/StringUtils.cs:82]: I found there's `CUnescape` extension method provided by Godot for strings, looking at its source code it seems to be a couple of replace calls which I think shouldn't be very expensive, and it also handles common escape sequences. I tested with the double backslash and a new line escape sequence and they're unescaped correctly in the output string. I think we could use this?

Comment (hhyyrylainen) [reply src/general/StringUtils.cs:82]: Well maybe. If it is literally just a couple of replaces, we could write our own (there's cost of marshaling data each time C++ code is called from C#, so it's likely faster to just go with a pure C# solution).

Comment (Kasterisk) [reply src/general/StringUtils.cs]: Added.

File: src/general/StringUtils.cs
Comment (hhyyrylainen) [inline src/general/StringUtils.cs:53]:
```suggestion
    /// <returns>A list of the substrings, starting from the beginning of the input string.</returns>
```

File: src/general/StringUtils.cs
Comment (hhyyrylainen) [inline src/general/StringUtils.cs:66]: This seems a bit too complex to be a one-liner...

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: src/gui_common/CustomRichTextLabel.cs
Comment (Kasterisk) [reply src/gui_common/CustomRichTextLabel.cs:247]: That could work.

File: src/general/StringUtils.cs
Comment (Kasterisk) [reply src/general/StringUtils.cs:152]: Well I assume the other sequences would do what they represent since newline, double quotes and backslash works correctly. I guess another static method could be added just for unescaping the whitespace controls?

Comment (hhyyrylainen) [reply src/general/StringUtils.cs:152]: Because I think we might use rich text for stuff like user input (in game chat for multiplayer if we add that), it's absolutely necessary that by *default* these bell and backspace escape codes are not allowed.

Comment (Kasterisk) [reply src/general/StringUtils.cs]: Removed some of the sequences.

File: src/general/StringUtils.cs
Comment (hhyyrylainen) [inline src/general/StringUtils.cs:72]: Why is this check here? Wouldn't it be safer to assume that if there is no ending `"` then rest of the string is within quotes?
Also this doesn't match the design, as we are already going to iterate over everything, I see this as pretty harmful to run a second loop to find the next `"` character.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: src/general/StringUtils.cs
Comment (Kasterisk) [reply src/general/StringUtils.cs:72]: Yeah I think that makes sense, though this needs mentioning in the docs that it doesn't always consider enclosing quotes.

**Review by hhyyrylainen — APPROVED** [review]
I think this has been tweaked enough and is fine to merge.

**Pull request merged**
